File: programmers/lv3/베스트앨범.py
def solution(genres, plays):
    answer = []
    
    dic_cnt = {}
    dic_li = {}
    for i in range(len(genres)):
        dic_cnt.setdefault(genres[i], 0)
        dic_li.setdefault(genres[i], [])
        dic_cnt[genres[i]] += plays[i]
        dic_li[genres[i]].append((plays[i], i))
    
    for key in dic_li:
        dic_li[key].sort(key=lambda x: (-x[0], x[1]))
    
    for key, _ in sorted(dic_cnt.items(), key=lambda x: -x[1]):
        answer.append(dic_li[key][0][1])
        if len(dic_li[key]) > 1:
            answer.append(dic_li[key][1][1])
        
    
    return answer

File: programmers/lv3/베스트앨범.java
import java.util.*;
import java.util.stream.IntStream;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> genreScores = new HashMap<>();
        Map<Integer, Integer> songScores = new HashMap<>();
        Map<String, Integer> genreCount = new HashMap<>();

        for (int i = 0; i < genres.length; i++) {
            String genre = genres[i];
            int play = plays[i];

            songScores.put(i, play);
            if (!genreScores.containsKey(genre)) {
                genreScores.put(genre, 0);
                genreCount.put(genre, 0);
            }
            genreScores.replace(genre, genreScores.get(genre) + play);
        }

        ArrayList<Integer> list = new ArrayList<>();

        IntStream.range(0, genres.length)
                .boxed()
                .sorted(Comparator.comparing(i -> -genreScores.get(genres[(Integer) i]))
                        .thenComparing(i -> -songScores.get((Integer) i))
                        .thenComparing(i -> (Integer) i))
                .mapToInt(i -> i)
                .forEach(i -> {
                    String genre = genres[i];
                    if (genreCount.get(genre) < 2) {
                        genreCount.replace(genre, genreCount.get(genre) + 1);
                        list.add(i);
                    }
                });

        return list.stream().mapToInt(i -> i).toArray();
    }
}